File: python/알고리즘/baek/브루트포스/15961_회전초밥.py
N, d, k, c = map(int, input().split())

rail = [int(input()) for _ in range(N)]

selected = [0] * 3001

# 현재 선택한 초밥의 종류
kinds = 0

# 초밥의 종류를 최대한 많이 선택한 경우
maximum_kinds = 0

for i in range(0,k):
    if selected[rail[i]] == 0:
        kinds += 1
    selected[rail[i]] += 1

maximum_kinds = kinds + 1 if selected[c] == 0 else kinds

for i in range(N):
    # 앞쪽꺼 하나 제외, 뒤쪽꺼 하나 더하고
    front = i
    end = i + k
    
    if end >= N:
        end -= N
    
    selected[rail[front]] -= 1
    
    if selected[rail[front]] == 0:
        kinds -= 1
    
    if selected[rail[end]] == 0:
        kinds += 1
    selected[rail[end]] += 1
    
    if selected[c] == 0:
        maximum_kinds = max(maximum_kinds, kinds + 1)
    else:
        maximum_kinds = max(maximum_kinds, kinds)

print(maximum_kinds)

File: python/알고리즘/baek/브루트포스/15961_회전초밥.c
#include <stdio.h>

int N, D, K, C;
int A[3000000+10];

void InputData(void){
    scanf("%d %d %d %d", &N, &D, &K, &C);
    for (int i=0; i<N ; i++){
        scanf("%d", &A[i]);
    }
}

int main(void){

    InputData();

    int selected_list[10000];
    int kinds = 0;
    int maximum_kinds = 0;
    for (int i = 0; i < K; i++) {
        if (selected_list[A[i]] == 0) {
            kinds++;
        }
        selected_list[A[i]]++;
    }

    if (selected_list[C] == 0) {
        maximum_kinds = kinds + 1;
    } else {
        maximum_kinds = kinds;
    }

    for (int i = 0; i < N; i++) {
        int front = i;
        int end = i + K;

        if (end >= N) {
            end -= N;
        }
        
        selected_list[A[front]]--;
        if (selected_list[A[front]] == 0) {
            kinds--;
        }

        if (selected_list[A[end]] == 0 ) {
            kinds++;
        }
        selected_list[A[end]]++;

        if (selected_list[C] == 0) {
            if (maximum_kinds < kinds + 1) {
                maximum_kinds = kinds + 1;
            }
        } else {
            if (maximum_kinds < kinds) {
                maximum_kinds = kinds;
            }
        }
    }

    printf("%d\n", maximum_kinds);
    return 0;
}